pkg/trace/metrics/timing/timing.go

```diff
@@ -0,0 +1,143 @@
+// Package timing is used to aggregate timing calls within hotpaths to avoid using
+// repeated statsd calls. The package has a default set that reports at 10 second
+// intervals and can be used directly. If a different behaviour or reporting pattern
+// is desired, a custom Set may be created.
+package timing
+
+import (
+	"sync"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/trace/atomic"
+	"github.com/DataDog/datadog-agent/pkg/trace/metrics"
+)
+
+// AutoreportInterval specifies the interval at which the default set reports.
+const AutoreportInterval = 10 * time.Second
+
+var (
+	defaultSet = NewSet()
+	stopReport = defaultSet.Autoreport(AutoreportInterval)
+)
+
+// Since records the duration for the given metric name as time passed since start.
+// It uses the default set which is reported at 10 second intervals.
+func Since(name string, start time.Time) { defaultSet.Since(name, start) }
+
+// Flush reports the default set. It can be useful to call when the program
+// exits prematurely.
+func Flush() { stopReport() }
+
+// Set represents a set of metrics that can be used for timing. Use NewSet
+// to create one and Report (or Autoreport) to submit metrics. It is safe for
+// concurrent use.
+type Set struct {
+	mu sync.RWMutex        // guards c
+	c  map[string]*counter // maps names to their aggregates
+}
+
+// NewSet returns a new Set, optionally initialized with the given metric names.
+func NewSet(names ...string) *Set {
```

Comment:
If the lazy-initialization works correctly, is there any reason to pre-declare metrics here? We should probably just get rid of this function as I don't see a reason to use it.

Reply:
I'm thinking this will help pre-initialize the map to avoid _any_ locking at all. It can't hurt to have it. It might also serve as a nice summary of all the metrics in the app, if we get to use this throughout the project. When we do move this to the `metrics` package, maybe we can have something like:

```go
var defaultSet = NewSet(
    "datadog.trace_agent.heartbeat",
    "datadog.trace_agent.started",
    // ... all agent metrics
)
```

It will be a nice opportunity for a full summary and even potentially documenting them (that's why the initial `panic` I added was there - to make sure that everything is declared in the constructor, for clarity reasons).

Getting rid of it however is also not a bad idea. I like the thought of having "read to use as is" identifiers which are simply declared as `var set timing.Set`. But I feel that for now, as an initial version, this should be enough. It would complicate `Add` even more to cover for a second case where the map is `nil` initially. I feel `Add` is already more complicated than it should be.

We can think about this more as we move forward. We are lucky in the sense that we don't have to maintain backwards compatibility here and it's fine to change these things later. What do you think?

Reply:
> It might also serve as a nice summary of all the metrics in the app, if we get to use this throughout the project.

Yeah but if there's no enforcing of requiring metrics to be declared, inevitably the only way to see all the metrics will be to grep for all calls to `timing.Since`. So this pre-declared set is almost certainly going to get out of sync with what metrics are actually being sent, making it more confusing that useful. 

There's that adage that I think applies very well here: *perfection is attained not when there is nothing more to add, but when there is nothing more to remove*